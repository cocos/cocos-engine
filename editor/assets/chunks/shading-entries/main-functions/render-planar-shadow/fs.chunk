in highp float v_dist;

layout(location = 0) out vec4 fragColorX;

void main () {
  // better performance than SurfacesFragmentModifyBaseColorAndTransparency
#ifdef CC_SURFACES_FRAGMENT_ALPHA_CLIP_ONLY
  SurfacesFragmentAlphaClipOnly();
#endif

  if(v_dist < 0.0)
    discard;
  fragColorX = cc_shadowColor;
}
