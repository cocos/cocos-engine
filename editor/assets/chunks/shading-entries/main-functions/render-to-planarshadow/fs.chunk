in highp float v_dist;

layout(location = 0) out vec4 fragColorX;

void main () {
  if(v_dist < 0.0)
    discard;
  fragColorX = cc_shadowColor;
}
