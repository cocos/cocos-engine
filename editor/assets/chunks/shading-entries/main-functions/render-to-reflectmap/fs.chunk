layout(location = 0) out vec4 fragColorX;

#if CC_PIPELINE_TYPE == CC_PIPELINE_TYPE_DEFERRED && !CC_FORCE_FORWARD_SHADING
  void main()  { fragColorX = vec4(0.0, 1.0, 0.0, 1.0); }
#else  
  void main()  {
  #if CC_DISABLE_STRUCTURE_IN_FRAGMENT_SHADER
    float NoL = dot(-cc_mainLitDir.xyz, FSInput_worldNormal.xyz); //trigger ubo binding
    vec4 color = SurfacesFragmentModifyBaseColorAndTransparency();
  #else
    // Surface
    SurfacesMaterialData surfaceData;
    CCSurfacesFragmentGetMaterialData(surfaceData);
    
    // Shadow parameters
    vec2 shadowBias = vec2(0.0);
    #if CC_RECEIVE_SHADOW
      shadowBias = FSInput_shadowBias;
    #endif
  
        
    // Fog factor
    #if !CC_FORWARD_ADD
      float fogFactor = 1.0;
      #if CC_USE_FOG != CC_FOG_NONE
        #if !CC_USE_ACCURATE_FOG
          fogFactor = FSInput_fogFactor;
        #else
          CC_TRANSFER_FOG_BASE(vec4(FSInput_worldPos, 1.0), fogFactor);
        #endif
      #endif
    #endif
  
  
    // Lighting
    LightingResult lightingResult;
    CCSurfacesLighting(lightingResult, surfaceData, shadowBias);
  
    // Shading
    vec4 color = CCSurfacesShading(surfaceData, lightingResult);
  
    #if !CC_FORWARD_ADD
      CC_APPLY_FOG_BASE(color, fogFactor);
    #endif
  #endif
  
    // Color output (RGBE)
    fragColorX = packRGBE(color.rgb);
  }
#endif
