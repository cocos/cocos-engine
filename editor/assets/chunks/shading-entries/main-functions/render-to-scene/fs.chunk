#if (CC_PIPELINE_TYPE == CC_PIPELINE_TYPE_FORWARD || CC_FORCE_FORWARD_SHADING)
  #include <shading-entries/main-functions/render-to-scene/pipeline/forward-fs>
#elif CC_PIPELINE_TYPE == CC_PIPELINE_TYPE_DEFERRED
  #include <shading-entries/main-functions/render-to-scene/pipeline/deferred-fs>
#endif  
