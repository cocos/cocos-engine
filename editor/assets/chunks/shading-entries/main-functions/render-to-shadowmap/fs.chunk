in highp vec2 v_clip_depth;

layout(location = 0) out vec4 fragColorX;

void main () {
  // better performance than SurfacesFragmentModifyBaseColorAndTransparency
  SurfacesFragmentAlphaClipOnly();
  

  highp float clipDepth = v_clip_depth.x / v_clip_depth.y * 0.5 + 0.5;
  // spot use linear
  if(cc_shadowLPNNInfo.x > EPSILON && cc_shadowLPNNInfo.x < 1.999999) {
    // enabled linear depth
    #if CC_SHADOWMAP_USE_LINEAR_DEPTH
      clipDepth = CCGetLinearDepth(FSInput_worldPos.xyz);
    #endif
  }

  #if CC_SHADOWMAP_FORMAT == SHADOWMAP_FORMAT_RGBA8
    fragColorX = packDepthToRGBA(clipDepth);
  #else
    fragColorX = vec4(clipDepth, 1.0, 1.0, 1.0);
  #endif
}
