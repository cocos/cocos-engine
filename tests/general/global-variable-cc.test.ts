/* eslint-disable @typescript-eslint/no-require-imports */

import fs from 'fs-extra';
import ps from 'path';

async function summarizeFilesInDirectory (directory: string) {
    return (await fs.readdir(directory)).map((file) => ps.join(directory, file));
}

declare global {
    const cc: object;
    const _cc: object;
}

test('global variables are sealed', async () => {
    const files = await summarizeFilesInDirectory(ps.join(__dirname, '..', '..', 'exports'));
    for (const file of files) {
        require(file);
    }

    let globalVariables: string[] = [];
    const visited: Set<object> = new Set();
    const getMembersRecursive = (object: {}, prefix: string) => {
        const memberNames = Object.getOwnPropertyNames(object);
        for (const memberName of memberNames) {
            const member = object[memberName];
            if (visited.has(member)) {
                continue;
            } else {
                visited.add(member);
            }
            const path = `${prefix}.${memberName}`;
            globalVariables.push(path);
            // TODO: Recursive
            // if (member === null) {
            //     console.warn(`${path} is \`null\`, is it intended?`);
            // }
            // else if (typeof member === 'object' && member.constructor === Object) {
            //     getMembersRecursive(member, path);
            // }
        }
    };
    // Global variable `cc`
    // getMembersRecursive(cc, 'cc'); // opted out for performance reasons (we are not doing this recursively anyways)
    globalVariables = Object.getOwnPropertyNames(cc).map((name) => `cc.${name}`);

    const sealedGlobalVariablesFile = ps.join(__dirname, 'sealed-global-variables.json');
    const output = false;
    if (output) {
        await fs.writeFile(sealedGlobalVariablesFile, JSON.stringify({
            thisFile: {
                stamp: new Date().toISOString(),
            },
            variables: globalVariables.sort(),
        }, undefined, 4));
    } else {
        const sealedGlobalVariables = (await fs.readJson(sealedGlobalVariablesFile)).variables as string[];
        expect(globalVariables.length === sealedGlobalVariables.length && globalVariables.every((v) => sealedGlobalVariables.includes(v))).toBeTruthy();
    }

}, 50000);
