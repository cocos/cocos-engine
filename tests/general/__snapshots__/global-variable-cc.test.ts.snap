// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`global variables are sealed 1`] = `
"cc.AffineTransform
cc.Ambient
cc.AmbientInfo
cc.AnimCurve
cc.Animation
cc.AnimationClip
cc.AnimationComponent
cc.AnimationManager
cc.AnimationState
cc.Asset
cc.AssetLibrary
cc.AssetManager
cc.AudioClip
cc.AudioSourceComponent
cc.BitMask
cc.BitmapFont
cc.Boolean
cc.BufferAsset
cc.Button
cc.CCBoolean
cc.CCFloat
cc.CCInteger
cc.CCString
cc.Camera
cc.CameraVisFlags
cc.Canvas
cc.Class
cc.Color
cc.Component
cc.ComponentModifier
cc.ConstantForce
cc.CubicSplineNumberValue
cc.CubicSplineQuatValue
cc.CubicSplineVec2Value
cc.CubicSplineVec3Value
cc.CubicSplineVec4Value
cc.CurveValueAdapter
cc.Director
cc.ENGINE_VERSION
cc.EffectAsset
cc.EmptyDevice
cc.Enum
cc.Event
cc.EventTarget
cc.Float
cc.Fog
cc.Font
cc.GFXAddress
cc.GFXAttributeName
cc.GFXBlendFactor
cc.GFXBlendOp
cc.GFXBuffer
cc.GFXBufferAccessBit
cc.GFXBufferFlagBit
cc.GFXBufferTextureCopy
cc.GFXBufferUsageBit
cc.GFXClearFlag
cc.GFXColor
cc.GFXColorMask
cc.GFXCommandBuffer
cc.GFXCommandBufferType
cc.GFXComparisonFunc
cc.GFXCullMode
cc.GFXDescriptorType
cc.GFXDevice
cc.GFXDynamicStateFlagBit
cc.GFXExtent
cc.GFXFilter
cc.GFXFormat
cc.GFXFormatInfo
cc.GFXFormatInfos
cc.GFXFormatSize
cc.GFXFormatSurfaceSize
cc.GFXFormatType
cc.GFXFramebuffer
cc.GFXGetTypeSize
cc.GFXInputAssembler
cc.GFXLoadOp
cc.GFXMemoryStatus
cc.GFXMemoryUsageBit
cc.GFXObject
cc.GFXObjectType
cc.GFXOffset
cc.GFXPipelineBindPoint
cc.GFXPipelineState
cc.GFXPolygonMode
cc.GFXPrimitiveMode
cc.GFXQueue
cc.GFXQueueType
cc.GFXRect
cc.GFXRenderPass
cc.GFXSampleCount
cc.GFXSampler
cc.GFXShadeModel
cc.GFXShader
cc.GFXShaderStageFlagBit
cc.GFXStencilFace
cc.GFXStencilOp
cc.GFXStoreOp
cc.GFXTexture
cc.GFXTextureCopy
cc.GFXTextureFlagBit
cc.GFXTextureLayout
cc.GFXTextureSubres
cc.GFXTextureType
cc.GFXTextureUsageBit
cc.GFXType
cc.GFXViewport
cc.GFX_MAX_ATTACHMENTS
cc.GFX_MAX_BUFFER_BINDINGS
cc.GFX_MAX_TEXTURE_UNITS
cc.GFX_MAX_VERTEX_ATTRIBUTES
cc.Game
cc.Graphics
cc.HierachyModifier
cc.ImageAsset
cc.Integer
cc.JsonAsset
cc.Label
cc.LabelAtlas
cc.LabelOutline
cc.Layers
cc.Layout
cc.LoadingItems
cc.Mask
cc.Mat3
cc.Mat4
cc.Material
cc.Mesh
cc.Node
cc.NodeEventProcessor
cc.NodePool
cc.Object
cc.PageView
cc.PageViewIndicator
cc.ParticleAsset
cc.ParticleUtils
cc.PhysicsMaterial
cc.PhysicsRayResult
cc.PhysicsSystem
cc.Pipeline
cc.Prefab
cc.Primitive
cc.PrivateNode
cc.ProgressBar
cc.Quat
cc.RatioSampler
cc.Rect
cc.RenderFlow
cc.RenderPassStage
cc.RenderPipeline
cc.RenderStage
cc.RenderTexture
cc.RenderableComponent
cc.ResolutionPolicy
cc.RichText
cc.Root
cc.SafeArea
cc.Scene
cc.SceneAsset
cc.SceneGlobals
cc.Scheduler
cc.ScrollBar
cc.ScrollView
cc.Shadows
cc.ShadowsInfo
cc.SimpleTexture
cc.Size
cc.Skeleton
cc.Skybox
cc.SkyboxInfo
cc.Slider
cc.Sprite
cc.SpriteAtlas
cc.SpriteFrame
cc.String
cc.SubContextView
cc.SystemEvent
cc.SystemEventType
cc.TTFFont
cc.TextAsset
cc.Texture2D
cc.TextureBase
cc.TextureCube
cc.Toggle
cc.ToggleContainer
cc.Touch
cc.Tween
cc.UI
cc.UIMeshRenderer
cc.UniformCurveValueAdapter
cc.ValueType
cc.Vec2
cc.Vec3
cc.Vec4
cc.ViewGroup
cc.VisibilityFlags
cc.WebGL2Device
cc.WebGLDevice
cc.Widget
cc._BaseNode
cc._JavaScript
cc._MissingScript
cc._Prefab
cc._PrefabInfo
cc._RF
cc._Script
cc._Test
cc._TypeScript
cc._componentMenuItems
cc._decorator
cc._global
cc._throw
cc._widgetManager
cc.assert
cc.assertID
cc.assetManager
cc.bezier
cc.bezierByTime
cc.builtinResMgr
cc.color
cc.debug
cc.deserialize
cc.director
cc.easing
cc.error
cc.errorID
cc.find
cc.game
cc.geometry
cc.getTypedArrayConstructor
cc.gfx
cc.instantiate
cc.internal
cc.isCustomTargetModifier
cc.isElementModifier
cc.isPropertyModifier
cc.isValid
cc.js
cc.loader
cc.log
cc.logID
cc.macro
cc.mat4
cc.math
cc.misc
cc.path
cc.physics
cc.primitives
cc.profiler
cc.programLib
cc.quat
cc.rect
cc.renderer
cc.resources
cc.sampleAnimationCurve
cc.screen
cc.size
cc.sys
cc.systemEvent
cc.tween
cc.tweenUtil
cc.url
cc.utils
cc.v2
cc.v3
cc.v4
cc.view
cc.visibleRect
cc.vmath
cc.warn
cc.warnID
cc.winSize"
`;
